hare3/hare_test.go

```diff
@@ -547,24 +579,52 @@ func (t *testTracer) waitSent() *Message {
 	return waitForChan(t.TB, t.sent, 10*time.Second, "no message")
 }
 
-func (*testTracer) OnStart(types.LayerID) {}
+func (t *testTracer) OnStart(l types.LayerID) {
+	t.started.Store(l.Uint32())
+	t.logger.Info("started", zap.Uint32("layer", l.Uint32()))
+}
 
 func (t *testTracer) OnStop(lid types.LayerID) {
+	t.logger.Info("stopped", zap.Uint32("layer", lid.Uint32()))
 	select {
 	case t.stopped <- lid:
 	default:
 	}
 }
 
 func (t *testTracer) OnActive(el []*types.HareEligibility) {
+	count := len(el)
+	if count == 0 {
+		t.logger.Info("no active eligibilities")
+	} else {
+		t.logger.Info(
+			"active eligibilities",
+			zap.Array("per signer", zapcore.ArrayMarshalerFunc(func(ae zapcore.ArrayEncoder) error {
+				for _, elig := range el {
+					if elig != nil {
+						ae.AppendUint16(elig.Count)
+					} else {
+						ae.AppendString("nil")
+					}
+
+				}
+				return nil
+			})),
+		)
+
+	}
```

Comment:
I think you don't need to handle `len(el) == 0` differently:

```suggestion
	t.logger.Info(
		"active eligibilities",
		zap.Array("per signer", zapcore.ArrayMarshalerFunc(func(ae zapcore.ArrayEncoder) error {
			for _, elig := range el {
				if elig != nil {
					ae.AppendUint16(elig.Count)
				} else {
					ae.AppendString("nil")
				}
			}
			return nil
		})),
	)
```

This would then just print `"m": "active eligibilities", "per signer": []`